oak_core/src/timer.rs

```diff
@@ -0,0 +1,191 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+/// Read the timestamp counter register from the CPU.
+///
+/// The timestamp counter is incremented once every CPU clock cycle, and how it behaves across power
+/// management events (e.g. change in P-state) is CPU-depdendent.
+///
+/// See RDTSC in AMD64 Architecture Programmer's Manual, Volume 3 for more details.
+pub fn rdtsc() -> u64 {
+    let mut edx: u64;
+    let mut eax: u64;
+
+    // Safety: `rdtsc` does not reference memory at all, so it's safe.
+    unsafe {
+        core::arch::asm!(
+            "rdtsc",
+            out("edx") edx,
+            out("eax") eax,
+            options(nomem, nostack)
+        )
+    }
+
+    edx << 32 | eax
+}
+
+/// Measures the number of clock cycles between `new()` and `elapsed()`. This measurement only makes
+/// sense if the process is running on the same CPU, as different CPUs will have different tick
+/// counter values.
+pub struct Timer {
+    start: u64,
+}
+
+impl Timer {
+    /// Constructs a new timer, recording the current tick counter value.
+    pub fn new() -> Self {
+        Self { start: rdtsc() }
+    }
+
+    /// Returns the approximate number of clock cycles it took to execute `func`.
+    pub fn timed<F>(func: F) -> u64
+    where
+        F: FnOnce(),
+    {
+        let timer = Timer::new();
+        func();
+        timer.elapsed()
+    }
+
+    /// Returns the approximate number of clock cycles elapsed since the construction of the
+    /// `Timer`.
+    pub fn elapsed(&self) -> u64 {
+        let stop = rdtsc();
+        stop - self.start
+    }
+}
+
+/// Basic histogram to compute percentiles over some measurement.
+/// N is the number of elements kept in memory; it must be larger than 0.
+/// The array is allocated statically; Histogram does not use heap allocation.
+#[derive(Debug, PartialEq)]
+pub struct Histogram<const N: usize> {
```

Comment:
nit: I don't think that `Histogram` is an accurate name here. Based on the name I initially assumed this would count the number of values falling in different ranges, so it took me quite some time to figure out what the code was doing. Perhaps something like `SampleStore` would be better.